pkg/runtime/k0s/runtime.go

```diff
@@ -0,0 +1,197 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k0s
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/sealerio/sealer/common"
+	"github.com/sealerio/sealer/pkg/registry"
+	"github.com/sealerio/sealer/pkg/runtime"
+	"github.com/sealerio/sealer/pkg/runtime/k0s/k0sctl"
+	v2 "github.com/sealerio/sealer/types/api/v2"
+	utilsnet "github.com/sealerio/sealer/utils/net"
+	"github.com/sealerio/sealer/utils/platform"
+	"github.com/sealerio/sealer/utils/ssh"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/sirupsen/logrus"
+)
+
+type Config struct {
+	Vlog      int
+	RegConfig *registry.Config
+	// 	clusterFileK0sConfig is k0sctl cluster config file
+	K0sConfig *k0sctl.K0sConfig
+}
+
+// Runtime struct is the runtime interface for k0s
+type Runtime struct {
+	*sync.Mutex
+	// cluster is sealer clusterFile
+	cluster *v2.Cluster
+	*Config
+}
+
+func (k *Runtime) Init() error {
+	return k.init()
+}
+
+func (k *Runtime) Upgrade() error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (k *Runtime) Reset() error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (k *Runtime) JoinMasters(newMastersIPList []net.IP) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (k *Runtime) JoinNodes(newNodesIPList []net.IP) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (k *Runtime) DeleteMasters(mastersIPList []net.IP) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (k *Runtime) DeleteNodes(nodesIPList []net.IP) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (k *Runtime) GetClusterMetadata() (*runtime.Metadata, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (k *Runtime) UpdateCert(certs []string) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+// NewK0sRuntime arg "clusterConfig" is the k0s config file under etc/${ant_name.yaml}, runtime need read k0s config from it
+// Mount image is required before new Runtime.
+func NewK0sRuntime(cluster *v2.Cluster) (runtime.Interface, error) {
+	return newK0sRuntime(cluster)
+}
+
+func newK0sRuntime(cluster *v2.Cluster) (runtime.Interface, error) {
+	k := &Runtime{
+		cluster: cluster,
+		Config:  &Config{},
+	}
+
+	k.Config.RegConfig = registry.GetConfig(k.getImageMountDir(), k.cluster.GetMaster0IP())
+
+	if err := k.checkList(); err != nil {
+		return nil, err
+	}
+
+	if logrus.GetLevel() == logrus.DebugLevel {
+		k.Vlog = 6
+	}
+	return k, nil
+}
+
+// checkList do a simple check for cluster spec Hosts which can not be empty.
+func (k *Runtime) checkList() error {
+	if len(k.cluster.Spec.Hosts) == 0 {
+		return fmt.Errorf("hosts spec cannot be empty")
+	}
+	if k.cluster.GetMaster0IP() == nil {
+		return fmt.Errorf("controller hosts ip cannot be empty")
```

Comment:
the error message seems to be not so readable. Just use`master0 IP cannot be empty`. Since no one could recognize what does controller mean.